### ISSUE TITLE
[question] How to build & install a package with a requirement has binaries only with no source code?

### ISSUE THREAD

**Issue by Bassem-Ramzy** [conan-io/conan#10736]
Hi,
I have three projects as following:
1. Open-source exe (myexe) that requires a library (lib1). Its package myexe/0.1 has source code exported.
2. Open-source library (lib1) that requires a library (lib2). Its lib1/0.1 package has source code exported.
3. _Closed-source_ library (lib2). Its package lib2/0.1 has _only_ lib2.lib file with no source code, created using `conan export-pkg . --build-folder path/to/my/local/build/folder`.

On _client's_ environment, when they need to install package while building myexe using conan, they do:
```CMD
conan install myexe/0.1@ --build
```
But, `package()` in lib2 is generated empty as it doesn't find the binaries as there is no path/to/my/local/build/folder.
Also, in case lib2 was already installed on the local cache (e.g. alone previously using `conan install lib2/0.1@`), when using the above command with `--build`, lib2 packages is forced to build, and conan tries to recreate the package binaries on the local cache as mentioned above, and binaries are deleted from the package.

So, how to install and build myexe on the client's machine? 

N.B. having lib2 packaged using `export-pkg` is done as per #10682.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @Bassem-Ramzy 

I think you are looking for the ``build_policy = "never"``, which allows skipping the build of export-pkg created packages. Please have a look at https://docs.conan.io/en/latest/mastering/policies.html?highlight=build_policy and tell us if this is what you need. Thanks

**Comment by Bassem-Ramzy** [comment]
Thanks James for your prompt reply, as always.

When adding:
```PYTHON
    build_policy = "never"
```
I'm getting an errors that don't happen without having "never" set:
```CMD
ERROR: No package matching 'xtdc4_driver-bin' pattern found.
ERROR: Trying to modify locked package xtdc4_driver-bin/0.0.1
```
Full output:
```CMD
>conan export-pkg . --build-folder=..\..\build\bfvs --force
Exporting package recipe
xtdc4_driver-bin/0.0.1 exports: Copied 1 '.py' file: CronoConanBase.py
xtdc4_driver-bin/0.0.1: Calling export_sources()
xtdc4_driver-bin/0.0.1: A new conanfile.py version was exported
xtdc4_driver-bin/0.0.1: Folder: C:\Users\basse\.conan\data\xtdc4_driver-bin\0.0.1\_\_\export
xtdc4_driver-bin/0.0.1: Exported revision: 7723193112aac890f7c0cc7f53c8bb28
ERROR: No package matching 'xtdc4_driver-bin' pattern found.
Packaging to 0469c0b660282e8e8f1d59fdfff0ed13eaf83f14
xtdc4_driver-bin/0.0.1: Generating the package
xtdc4_driver-bin/0.0.1: Package folder C:\Users\basse\.conan\data\xtdc4_driver-bin\0.0.1\_\_\package\0469c0b660282e8e8f1d59fdfff0ed13eaf83f14
xtdc4_driver-bin/0.0.1: Calling package()
xtdc4_driver-bin/0.0.1 package(): Packaged 1 '.dll' file: xtdc4_driver.dll
xtdc4_driver-bin/0.0.1 package(): Packaged 1 '.lib' file: xtdc4_driver.lib
xtdc4_driver-bin/0.0.1: Package '0469c0b660282e8e8f1d59fdfff0ed13eaf83f14' created
xtdc4_driver-bin/0.0.1: Created package revision fec979a129933bb10f2d80167ed85bf3
ERROR: Trying to modify locked package xtdc4_driver-bin/0.0.1
```

Any precautions needed?

**Comment by memsharded** [comment]
Uhm, that seems some unexpected side effect of the previous steps before the ``conan export-pkg``. It seems a previous install is generating a lockfile that is being automatically used? It is a bit confusing, apparently ``conan export-pkg`` shouldn't be using the lockfile unless explicitly passed as argument with ``--lockfile=xxxx``

If you could try to reproduce, detailing the commands before the export-pkg, that could help, maybe with the starting point of ``conan new hello/0.1 --template=cmake_lib`` recipe template?

**Comment by Bassem-Ramzy** [comment]
Well, I did the following:
```CMD
mkdir policy
cd policy
conan new hello/0.1 --template=cmake_lib
cmake .
cmake --build .
conan export-pkg . --build-folder=Debug
```
It worked with the generated recipe.

Then updated the recipe to be:
```PYTHON
from conans import ConanFile
from conan.tools.cmake import CMakeToolchain, CMake


class HelloConan(ConanFile):
    name = "hello"
    version = "0.1"

    # Optional metadata
    license = "<Put the package license here>"
    author = "<Put your name here> <And your email here>"
    url = "<Package recipe repository url here, for issues about the package>"
    description = "<Description of Hello here>"
    topics = ("<Put some tag here>", "<here>", "<and here>")

    # Binary configuration
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": False, "fPIC": True}
    build_policy = "never"
    
    # Sources are located in the same place as this recipe, copy them to the recipe
    exports_sources = "CMakeLists.txt", "src/*"

    def config_options(self):
        if self.settings.os == "Windows":
            del self.options.fPIC

    def package(self):
        self.copy("*.lib")

    def package_info(self):
        self.cpp_info.libs = ["hello"]
```
But I got:
```CMD
> conan export-pkg . --build-folder=Debug --force
Exporting package recipe
hello/0.1 exports_sources: Copied 1 '.txt' file: CMakeLists.txt
hello/0.1 exports_sources: Copied 1 '.cpp' file: hello.cpp
hello/0.1 exports_sources: Copied 1 '.h' file: hello.h
hello/0.1: The stored package has not changed
hello/0.1: Exported revision: c688a4558fbed8b26db20960bd1fb887
ERROR: No package matching 'hello' pattern found.
Packaging to 3fb49604f9c2f729b85ba3115852006824e72cab
hello/0.1: Generating the package
hello/0.1: Package folder C:\Users\basse\.conan\data\hello\0.1\_\_\package\3fb49604f9c2f729b85ba3115852006824e72cab
hello/0.1: Calling package()
hello/0.1 package(): Packaged 1 '.lib' file: hello.lib
hello/0.1: Package '3fb49604f9c2f729b85ba3115852006824e72cab' created
hello/0.1: Created package revision 5470f106f890fa11734fd50b37962622
ERROR: Trying to modify locked package hello/0.1
```

**Comment by memsharded** [comment]
There are a couple of things that could be affecting:

- Don't modify the recipe and remove the ``layout()`` method, after it has already been used. The ``layout()`` method defines the location of things, you don't want to kind of use it for building, then remove it, results will  likely be inconsistent.
- For the local flow, you should be calling ``conan install`` first. Then, pass the generated toolchain to the ``cmake ... -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake``, because this is what defines that the current build is matching the used Conan settings.
- Also do not modify the ``build_policy`` after it has been exported for the first time, I'd recommend to change it before building.

In any case, it seems that this could be a bug in the ``build_policy = "never"``, we will investigate it, thanks for reporting!

**Comment by memsharded** [comment]
The bug was fixed in https://github.com/conan-io/conan/pull/10738, it will be released next 1.47.

In the meantime, the ``conan export-pkg`` with this policy will only work for new packages, with ``--force``. For building the graph, using ``--build=missing`` would be the recommended strategy, in general it is recommended anyway, as ``--build`` all is typically unnecessary and a waste of resources, and using ``--build=missing`` with the right versioning and package_id_mode will be faster and more efficient.
